client/windows/GUIClasses.cpp

```diff
@@ -1281,22 +1305,28 @@ std::string CHillFortWindow::getTextForSlot(SlotID slot)
 	return str;
 }
 
-int CHillFortWindow::getState(SlotID slot)
+CHillFortWindow::State CHillFortWindow::getState(SlotID slot)
 {
 	TResources myRes = LOCPLINT->cb->getResourceAmount();
 
-	if(hero->slotEmpty(slot))//no creature here
-		return -1;
+	if(hero->slotEmpty(slot))
+		return State::EMPTY;
 
 	UpgradeInfo info;
 	LOCPLINT->cb->fillUpgradeInfo(hero, slot, info);
-	if(!info.newID.size())//already upgraded
-		return 1;
+	if (!info.newID.size())
+	{
+		// new Hill Fort allows upgrades level 5 and below
+		if (hero->getStack(slot).type->getLevel() >= 5 && hero->getCreature(slot)->hasUpgrades())
```

Comment:
Why do you need check for level here?
If list of potential upgrades is empty, but creature has upgrades then upgrade is unavailable.
Also, hill fort level limits (and costs) are configurable, nothing stops modders from making hill fort with configuration different from hota
```suggestion
		// Hill Fort may limit level of upgradeable creatures, e.g. mini hill fort from hota
		if (hero->getCreature(slot)->hasUpgrades())
```